examples/typed-state/README.md

```diff
@@ -0,0 +1,263 @@
+
+While it will still work in pycharm, you will not get some of the beter auto-completion capabilities.
+
+## Defining typed state at the application level
+
+This code for this is in [application.py](application.py).
+
+First, define a pydantic model -- make it as recursive as you want. This will represent
+your entire state. In this case, we're going to have a transcript of a youtube video that
+was given by the user, as well as the social media post. The high-level is here -- the rest is
+in the code:
+
+```python
+class ApplicationState(BaseModel):
+    # Make these have defaults as they are only set in actions
+    transcript: Optional[str] = Field(
+        description="The full transcript of the YouTube video.", default=None
+    )
+    post: Optional[SocialMediaPost] = Field(
+        description="The generated social media post.", default=None
+    )
+```
+
+Note that this should exactly model your state -- we need to make things optional,
+as there are points in the state where the transcript/post will not have been assigned.
+
+Next, we add it in to the application object, both the initial value and the typing system.
+The typing system is what's responsible for managing the schema:
+
+```python
+app = (
+        ApplicationBuilder()
+        .with_actions(
+            ...
+        )
+        .with_transitions(
+            ...
+        )
+        .with_entrypoint(...)
+        .with_typing(PydanticTypingSystem(ApplicationState))
+        .with_state(ApplicationState())
+        .build()
+    )
+```
+
+That ensures that application and the application's state object are parameterized on the state type.
+
+To get the state, you can use `.data` on any state object returned by the application
+
+```python
+# just from the application
+print(app.state.data.transcript)
+
+# after execution
+_, _, state = app.run(halt_after=..., inputs=...)
+print(state.data.transcript)
+```
+
+## Defining/altering typed state at the action level
+
+This code for this is in [application.py](application.py).
+
+In addition to defining state centrally, we can define it at an action level.
+
+The code is simple, but the API is slightly different from standard Burr. Rather than
+using the immutable state-based API, we in-place mutate pydantic models. Don't worry, it's still immutable, you're just modifying a copy and returning it.
+
+In this case, we call to `@action.pydantic`, which tells which fields to read/write to from state. It derives the classes from the function annotations, although you can also pass it the
+pydantic classes as arguments to the decorator if you prefer.
+
+Note that the reads/writes have to be a subset of the state object. In this case we use the global `ApplicationState` object as described above, although it can use a subset/compatible set of fields (or, if you elect not to use centralized state, it just has to be compatible with upstream/downstream versions).
```

Comment:
```suggestion
Note that the reads/writes have to be a subset of your defined typed state object. In this case we use the global `ApplicationState` object as described above, although it can use a subset/compatible set of fields (or, if you elect not to use centralized state, it just has to be compatible with upstream/downstream versions).
```